hazelcast/src/main/java/com/hazelcast/sql/impl/plan/node/io/ReceiveSortMergePlanNode.java

```diff
@@ -0,0 +1,137 @@
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.plan.node.io;
+
+import com.hazelcast.internal.serialization.impl.SerializationUtil;
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.plan.node.PlanNodeSchema;
+import com.hazelcast.sql.impl.plan.node.ZeroInputPlanNode;
+import com.hazelcast.sql.impl.plan.node.PlanNodeVisitor;
+import com.hazelcast.sql.impl.type.QueryDataType;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+/**
+ * Physical node which receives from remote stripes and performs sort-merge.
+ */
+@SuppressWarnings("rawtypes")
+public class ReceiveSortMergePlanNode extends ZeroInputPlanNode implements EdgeAwarePlanNode {
+    /** Edge iD. */
+    private int edgeId;
+
+    /** Field types. */
+    private List<QueryDataType> fieldTypes;
+
+    /** Expressions to be used for sorting. */
+    private List<Expression> expressions;
+
+    /** Sort directions. */
+    private List<Boolean> ascs;
+
+    public ReceiveSortMergePlanNode() {
+        // No-op.
+    }
+
+    public ReceiveSortMergePlanNode(
+            int id,
+            int edgeId,
+            List<QueryDataType> fieldTypes,
+            List<Expression> expressions,
+            List<Boolean> ascs
+    ) {
+        super(id);
+
+        this.edgeId = edgeId;
+        this.fieldTypes = fieldTypes;
+        this.expressions = expressions;
+        this.ascs = ascs;
+    }
+
+    public List<Expression> getExpressions() {
+        return expressions;
+    }
+
+    public List<Boolean> getAscs() {
+        return ascs;
+    }
+
+    @Override
+    public int getEdgeId() {
+        return edgeId;
+    }
+
+    @Override
+    public boolean isSender() {
+        return false;
+    }
+
+    @Override
+    public void visit(PlanNodeVisitor visitor) {
+        visitor.onReceiveSortMergeNode(this);
+    }
+
+    @Override
+    public PlanNodeSchema getSchema0() {
+        return new PlanNodeSchema(fieldTypes);
+    }
+
+    @Override
+    public void writeData0(ObjectDataOutput out) throws IOException {
+        out.writeInt(edgeId);
+        SerializationUtil.writeList(fieldTypes, out);
+        out.writeObject(expressions);
+        out.writeObject(ascs);
+    }
+
+    @Override
+    public void readData0(ObjectDataInput in) throws IOException {
+        edgeId = in.readInt();
+        fieldTypes = SerializationUtil.readList(in);
+        expressions = in.readObject();
+        ascs = in.readObject();
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(id, edgeId, expressions, ascs);
```

Comment:
Is it ok that `fieldTypes` are not used in `equals/hashCode`?

Reply:
Fixed in https://github.com/hazelcast/hazelcast/pull/18013/commits/a4b6a6101fe6d4a13117ea2591d7fd17b0509e29